### PR TITLE
Refine build script for adding disable selected data types option

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
### Description

As title. 

### Motivation and Context

Now we have multiple data types that we want to disable for minimal build and to reduce binary size. may be worth adding an argument in the build script for specifying that.

Also for fp16 type stuff, it may be too restrict to disable that for all minimal build.

### CODE REVIEW

File: tools/ci_build/build.py [hunk 1]
```diff
@@ -631,6 +631,11 @@ def convert_arg_line_to_args(self, arg_line):
     )
     # Please note in our CMakeLists.txt this is already default on. But in this file we reverse it to default OFF.
     parser.add_argument("--disable_rtti", action="store_true", help="Disable RTTI (reduces binary size)")
+    parser.add_argument(
+        "--disable_types",
+        nargs="+",
+        default=[],
+        choices=["float8", "float16", "optional", "sparsetensor"], help="Disable selected data types (reduces binary size)")
```
Comment (YUNQIUGUO) [inline tools/ci_build/build.py:638]: Note: I didn't include abseil types here for now.

**Review by edgchen1 — DISMISSED** [review]

File: tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml [hunk 2]
```diff
@@ -247,11 +247,9 @@ jobs:
               --parallel \
               --skip_tests \
               --disable_ml_ops \
+              --disable_types sparsetensor float8 float16 optional \
```
[the remark this answers is not in the document]
Comment (YUNQIUGUO) [reply tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml:250]: right, updated.

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -139,6 +139,7 @@ option(onnxruntime_DISABLE_ML_OPS "Disable traditional ML ops" OFF)
 option(onnxruntime_DISABLE_SPARSE_TENSORS "Disable sparse tensors data types" OFF)
 option(onnxruntime_DISABLE_OPTIONAL_TYPE "Disable optional type" OFF)
 option(onnxruntime_DISABLE_FLOAT8_TYPES "Disable float 8 types" OFF)
+option(onnxruntime_DISABLE_FLOAT16_TYPES "Disable float 16 types" OFF)
```
Comment (skottmckay) [inline cmake/CMakeLists.txt:142]: float16 one seems a little misleading if it's only affecting part of MLAS given float16 is used in a lot of other places and we're not disabling any of those. 

what's the binary size saving of having this?

File: tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (edgchen1) [reply tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml:250]: the display name and comments at the top of this file say that "all optional features" are disabled for these steps. maybe "float8" and "float16" should be added to the other steps too?

File: cmake/CMakeLists.txt (hunk 3 above)
Comment (YUNQIUGUO) [reply cmake/CMakeLists.txt:142]: > the float16 one seems a little misleading if it's only affecting part of MLAS given float16 is used in a lot of other places and we're not disabling any of those.
> 
> what's the binary size saving of having this?

hm, okay that's true. excluding the fp16 support files in total saves about 20KB uncompressed.

how about using another name e.g. DISABLE_PARTIAL_MLAS_FLOAT16_TYPES?/better suggestions?

Comment (skottmckay) [reply cmake/CMakeLists.txt:142]: Possibly getting too specific for a top-level option in the ORT build.

Maybe we can add this to the list of potential reductions and do it in a more complete way if needed, as there's float16 as well as bfloat16 to consider.

I probably wouldn't exclude it for all minimal builds either as there's no way to enable it in a minimal build if you wanted to.

Comment (YUNQIUGUO) [reply cmake/CMakeLists.txt]: cool, updated.

**Review by skottmckay — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**